File: src/parser/parser.ts
import type {
	Block,
	Expr,
	ListLiteral,
	Literal,
	Print,
	RangeExpr,
	Stmt,
	Tangible,
	Variable,
} from "../ast";
import { TokenType, type Token } from "../lexer/lexer";

class ParseError extends Error {}

export class Parser {
	private current = 0;

	constructor(private tokens: Token[]) {}

	parse(): Stmt[] {
		const statements: Stmt[] = [];
		while (!this.isAtEnd()) {
			statements.push(this.declaration());
		}
		return statements;
	}

	private debug(label?: string) {
		console.log(label, {
			current: this.peek(),
			previous: this.tokens[this.current - 1],
			next: this.tokens[this.current + 1],
		});
	}

	private expression(): Expr {
		const expr = this.assignment();

		if (this.match(TokenType.LEFT_PAREN)) {
			return this.call(expr);
		}

		return expr;
	}

	private call(callee: Expr): Expr {
		if (this.match(TokenType.RIGHT_PAREN)) {
			return {
				type: "Call",
				callee,
				arguments: [],
			};
		}

		const args: Expr[] = [];
		do {
			args.push(this.expression());
		} while (this.match(TokenType.COMMA));
		this.consume(TokenType.RIGHT_PAREN, "Expect ')' to end function call.");

		return {
			type: "Call",
			callee,
			arguments: args,
		};
	}

	private assignment(): Expr {
		const expr = this.or();

		if (this.match(TokenType.INCREMENT)) {
			const plus = this.previous();
			const value = this.assignment();
			if (expr.type === "Variable") {
				return { type: "Increment", name: expr.token, value };
			}
			this.error(plus, "Invalid assignment target.");
		}
		if (this.match(TokenType.DECREMENT)) {
			const plus = this.previous();
			const value = this.assignment();
			if (expr.type === "Variable") {
				return { type: "Decrement", name: expr.token, value };
			}
			this.error(plus, "Invalid assignment target.");
		}

		if (this.match(TokenType.ASSIGN)) {
			const equals = this.previous();
			const value = this.assignment();
			if (expr.type === "Variable") {
				return { type: "Assign", name: expr.token, value };
			}
			this.error(equals, "Invalid assignment target.");
		}
		return expr;
	}

	private equality(): Expr {
		let expr = this.comparison();
		while (this.match(TokenType.NOT_EQUAL, TokenType.EQUAL)) {
			const operator = this.previous();
			const right = this.comparison();
			expr = { type: "Binary", left: expr, operator, right };
		}
		return expr;
	}

	private comparison(): Expr {
		let expr = this.term();
		while (
			this.match(
				TokenType.GREATER_THAN,
				TokenType.GREATER_EQUAL,
				TokenType.LESS_THAN,
				TokenType.LESS_EQUAL,
			)
		) {
			const operator = this.previous();
			const right = this.term();
			expr = { type: "Binary", left: expr, operator, right };
		}
		return expr;
	}

	private term(): Expr {
		let expr = this.factor();
		while (this.match(TokenType.MINUS, TokenType.PLUS)) {
			const operator = this.previous();
			const right = this.factor();
			expr = { type: "Binary", left: expr, operator, right };
		}
		return expr;
	}

	private factor(): Expr {
		let expr = this.unary();
		while (this.match(TokenType.SLASH, TokenType.STAR)) {
			const operator = this.previous();
			const right = this.unary();
			expr = { type: "Binary", left: expr, operator, right };
		}
		return expr;
	}

	private unary(): Expr {
		if (this.match(TokenType.BANG, TokenType.MINUS)) {
			const operator = this.previous();
			const right = this.unary();
			return { type: "Unary", operator, right };
		}
		return this.primary();
	}

	private declaration(): Stmt {
		try {
			if (this.match(TokenType.MUT)) return this.mutDeclaration();
			if (this.match(TokenType.LET)) return this.letDeclaration();
			if (this.match(TokenType.FUNC)) return this.function("function");
			if (this.match(TokenType.BLANK_LINE)) return { type: "BlankLine" };
			return this.statement();
		} catch (error) {
			this.synchronize();
			throw error;
		}
	}

	private statement(): Stmt {
		if (this.match(TokenType.PRINT)) return this.printStatement();
		if (this.match(TokenType.LEFT_BRACE)) return this.block();
		if (this.match(TokenType.IF)) return this.ifStatement();
		if (this.match(TokenType.WHILE)) return this.whileStatement();
		if (this.match(TokenType.FOR)) return this.forStatement();
		if (this.match(TokenType.RETURN)) return this.returnStatement();
		return this.expressionStatement();
		// TODO: immediately evaluated blocks
		// 	return { type: "Block", statements: this.block() };
	}

	private block(): Block {
		const statements: Stmt[] = [];
		while (!this.check(TokenType.RIGHT_BRACE) && !this.isAtEnd()) {
			statements.push(this.declaration());
		}
		this.consume(TokenType.RIGHT_BRACE, "Expect '}' after block.");
		return { type: "Block", statements };
	}

	private printStatement(): Print {
		const expression = this.expression();
		this.match(TokenType.SEMICOLON);
		return { type: "Print", expression };
	}

	private expressionStatement(): Stmt {
		const expr = this.expression();
		this.match(TokenType.SEMICOLON);
		return { type: "ExprStatement", expression: expr };
	}

	private letDeclaration(): Stmt {
		const name = this.consume(TokenType.IDENTIFIER, "Expect variable name.");
		let initializer: Expr | null = null;
		let variableType: Token | null = null;
		if (this.match(TokenType.COLON)) {
			variableType = this.consume(
				TokenType.IDENTIFIER,
				"Expect type after ':'.",
			);
		}
		if (this.match(TokenType.ASSIGN)) {
			initializer = this.expression();
			// semi-colon is optional
			if (this.match(TokenType.SEMICOLON)) {
			}
			return { type: "LetDecl", name, initializer, _staticType: variableType };
		}
		throw this.error(this.peek(), "Expect variable initializer.");
	}

	private mutDeclaration(): Stmt {
		const name = this.consume(TokenType.IDENTIFIER, "Expect variable name.");
		let initializer: Expr | null = null;
		let variableType: Token | null = null;
		if (this.match(TokenType.COLON)) {
			variableType = this.consume(
				TokenType.IDENTIFIER,
				"Expect type after ':'.",
			);
		}
		if (this.match(TokenType.ASSIGN)) {
			initializer = this.expression();
			// semi-colon is optional
			if (this.match(TokenType.SEMICOLON)) {
			}
			return { type: "MutDecl", name, initializer, _staticType: variableType };
		} else {
			throw this.error(this.peek(), "Expect variable initializer.");
		}
	}

	private ifStatement(): Stmt {
		const condition = this.expression();
		this.consume(TokenType.LEFT_BRACE, "Expect '{' after if condition.");
		const thenBranch: Stmt = this.block();
		let elseBranch: Stmt | null = null;
		if (this.match(TokenType.ELSE)) {
			this.consume(TokenType.LEFT_BRACE, "Expect '{' after else condition.");
			elseBranch = this.block();
		}
		return { type: "If", condition, thenBranch, elseBranch };
	}

	private whileStatement(): Stmt {
		const condition = this.expression();
		this.consume(TokenType.LEFT_BRACE, "Expect '{' after while condition.");
		const body = this.block();
		return { type: "While", condition, body };
	}

	private forStatement(): Stmt {
		const cursor = this.consume(
			TokenType.IDENTIFIER,
			"Expect cursor name after 'for'.",
		);
		this.consume(TokenType.IN, "Expect 'in' after cursor name.");
		const range = this.rangeExpression();
		this.consume(TokenType.LEFT_BRACE, "Expect '{' after 'for range'.");
		const body = this.block();
		return {
			type: "ForIn",
			cursor,
			range,
			body,
		};
	}

	private rangeExpression(): RangeExpr {
		const start = this.consume(
			TokenType.INTEGER,
			"Expect integer range after 'for'.",
		);
		this.consume(TokenType.RANGE_DOTS, "Expect integer range after 'for'.");
		const end = this.consume(
			TokenType.INTEGER,
			"Expect integer range after 'for'.",
		);
		return { type: "RangeExpr", start, end };
	}

	private function(kind: string): Stmt {
		const name = this.consume(TokenType.IDENTIFIER, `Expect ${kind} name.`);
		this.consume(TokenType.LEFT_PAREN, `Expect '(' after ${kind} name.`);
		const parameters: Token[] = [];
		if (!this.check(TokenType.RIGHT_PAREN)) {
			do {
				if (parameters.length >= 255) {
					this.error(this.peek(), "Can't have more than 255 parameters.");
				}
				parameters.push(
					this.consume(TokenType.IDENTIFIER, "Expect parameter name."),
				);
			} while (this.match(TokenType.COMMA));
		}
		this.consume(TokenType.RIGHT_PAREN, "Expect ')' after parameters.");
		this.consume(TokenType.LEFT_BRACE, `Expect '{' before ${kind} body.`);
		const body = this.block();
		return { type: "Function", name, params: parameters, body };
	}

	private returnStatement(): Stmt {
		const keyword = this.previous();
		let value = null;
		if (
			!this.check(TokenType.SEMICOLON) &&
			!this.check(TokenType.RIGHT_BRACE)
		) {
			value = this.expression();
		}
		return { type: "Return", keyword, value };
	}

	private or(): Expr {
		let expr = this.and();
		while (this.match(TokenType.OR)) {
			const operator = this.previous();
			const right = this.and();
			expr = { type: "Logical", left: expr, operator, right };
		}
		return expr;
	}

	private and(): Expr {
		let expr = this.equality();
		while (this.match(TokenType.AND)) {
			const operator = this.previous();
			const right = this.equality();
			expr = { type: "Logical", left: expr, operator, right };
		}
		return expr;
	}

	private primary(): Expr {
		if (this.match(TokenType.LEFT_PAREN)) {
			const expr = this.expression();
			this.consume(TokenType.RIGHT_PAREN, "Expect ')' after expression.");
			return { type: "Grouping", expression: expr };
		}
		return this.atom();
	}

	private atom(): Tangible {
		if (this.match(TokenType.FALSE)) {
			return { type: "Literal", value: false, token: this.previous() };
		}
		if (this.match(TokenType.TRUE)) {
			return { type: "Literal", value: true, token: this.previous() };
		}
		if (this.match(TokenType.INTEGER)) {
			const token = this.previous();
			return { type: "Literal", value: Number(token.lexeme), token };
		}
		// memo: this distinction probably doesn't matter in a JS runtime
		if (this.match(TokenType.DOUBLE)) {
			const token = this.previous();
			return { type: "Literal", value: parseFloat(token.lexeme), token };
		}
		if (this.match(TokenType.STRING)) {
			const token = this.previous();
			return {
				type: "Literal",
				value: token.lexeme,
				token,
			};
		}
		if (this.match(TokenType.IDENTIFIER)) {
			return { type: "Variable", token: this.previous() };
		}
		return this.list();
	}

	private list(): ListLiteral {
		this.consume(TokenType.LEFT_BRACKET, "Expect a '['.");
		let items: Tangible[] = [];
		if (!this.check(TokenType.RIGHT_BRACKET)) {
			do {
				items.push(this.atom());
			} while (this.match(TokenType.COMMA));
		}
		this.consume(TokenType.RIGHT_BRACKET, "Expect a ']' to close the list.");
		return { type: "List", items };
	}

	private match(...types: TokenType[]): boolean {
		for (const type of types) {
			if (this.check(type)) {
				this.advance();
				return true;
			}
		}
		return false;
	}

	private consume(type: TokenType, message: string): Token {
		if (this.check(type)) return this.advance();
		throw this.error(this.peek(), message);
	}

	private check(type: TokenType): boolean {
		if (this.isAtEnd()) return false;
		return this.peek().type === type;
	}

	private advance(): Token {
		if (!this.isAtEnd()) this.current++;
		return this.previous();
	}

	private isAtEnd(): boolean {
		return this.peek().type === TokenType.EOF;
	}

	private peek(): Token {
		const token = this.tokens[this.current];
		if (!token) throw new ParseError();
		return token;
	}

	private previous(): Token {
		const token = this.tokens[this.current - 1];
		if (!token) throw new ParseError();
		return token;
	}

	private error(token: Token, message: string): ParseError {
		// if (token.type == TokenType.EOF) {
		// 	console.log(`Syntax error on line ${token.line}, at end: ${message}`);
		// } else {
		// 	console.log(
		// 		`Syntax error on line ${token.line}, at ${token.lexeme}: ${message}`,
		// 	);
		// }
		return new ParseError(
			message + " at " + token.line + ":" + token.column + ".",
		);
	}

	private synchronize() {
		this.advance();
		while (!this.isAtEnd()) {
			// explicit semicolons may be an explicit statement boundary
			if (this.previous().type === TokenType.SEMICOLON) return;
			switch (this.peek().type) {
				case TokenType.FUNC:
				case TokenType.LET:
				case TokenType.MUT:
				case TokenType.FOR:
				case TokenType.IF:
				case TokenType.WHILE:
				case TokenType.RETURN:
					return;
			}
			this.advance();
		}
	}

	// private finishCall(callee: Expr): Expr {
	// 	const args: Expr[] = [];
	// 	if (!this.check(TokenType.RIGHT_PAREN)) {
	// 		do {
	// 			if (args.length >= 255) {
	// 				this.error(this.peek(), "Can't have more than 255 arguments.");
	// 			}
	// 			args.push(this.expression());
	// 		} while (this.match(TokenType.COMMA));
	// 	}
	// 	const paren = this.consume(
	// 		TokenType.RIGHT_PAREN,
	// 		"Expect ')' after arguments.",
	// 	);
	// 	return { type: "Call", callee, paren, arguments: args };
	// }
}

File: internal/ast/parser.go
package ast

import (
	"fmt"
	"slices"
	"strings"
)

type node struct {
	text       string
	start, end int
}

type parser struct {
	tokens []token
	index  int
}

func newParser(tokens []token) *parser {
	return &parser{
		tokens: tokens,
		index:  0,
	}
}

func (p *parser) parse() (*Program, error) {
	program := &Program{
		Imports:    []Import{},
		Statements: []Statement{},
	}

	// Parse imports first
	for p.match(use) {
		imp, err := p.parseImport()
		if err != nil {
			return nil, err
		}
		program.Imports = append(program.Imports, *imp)
	}

	// Parse statements
	for !p.isAtEnd() {
		if p.match(new_line) {
			continue
		}
		stmt, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		if stmt != nil {
			program.Statements = append(program.Statements, stmt)
		}
	}

	return program, nil
}

func (p *parser) parseImport() (*Import, error) {
	pathToken := p.consume(path, "Expected a module path after 'use'")

	var name string
	if p.match(as) {
		alias := p.consume(identifier, "Expected alias name after 'as'")
		name = alias.text
	} else {
		// Default alias is last part of path
		parts := strings.Split(pathToken.text, "/")
		if len(parts) == 1 {
			name = parts[0]
		} else {
			name = parts[len(parts)-1]
		}
		name = strings.ReplaceAll(name, "-", "_")
	}
	p.match(new_line)

	return &Import{
		Path: pathToken.text,
		Name: name,
	}, nil
}

func (p *parser) parseStatement() (Statement, error) {
	if p.match(let, mut) {
		return p.parseVariableDef()
	}
	if p.match(while_) {
		return p.whileLoop()
	}
	return p.assignment()
	// return p.expressionStatement()
}

func (p *parser) parseVariableDef() (Statement, error) {
	kind := p.tokens[p.index-1].kind
	name := p.consume(identifier, "Expected identifier after variable declaration")
	var declaredType DeclaredType
	if p.match(colon) {
		typeToken := p.consume(identifier, "Expected a type name after ':'")
		switch typeToken.text {
		case "Int":
			declaredType = IntType{}
		case "Float":
			declaredType = FloatType{}
		case "Str":
			declaredType = StringType{}
		case "Bool":
			declaredType = BooleanType{}
		default:
			declaredType = CustomType{
				Name: typeToken.text,
			}
		}
	}
	p.consume(equal, "Expected '=' after variable name")
	value, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	p.match(new_line)
	return &VariableDeclaration{
		Mutable: kind == mut,
		Name:    name.text,
		Value:   value,
		Type:    declaredType,
	}, nil
}

func (p *parser) whileLoop() (Statement, error) {
	condition, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	p.consume(left_brace, "Expected '{' after while condition")

	statements := []Statement{}
	for !p.check(right_brace) {
		stmt, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		statements = append(statements, stmt)
	}
	p.consume(right_brace, "Unclosed while loop")
	p.match(new_line)

	return &WhileLoop{
		Condition: condition,
		Body:      statements,
	}, nil
}

func (p *parser) assignment() (Statement, error) {
	expr, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	if p.match(equal, increment, decrement) {
		opToken := p.previous()
		value, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		p.match(new_line)

		switch opToken.kind {
		case increment:
			return &VariableAssignment{
				Operator: Assign,
				Target:   expr,
				Value: &BinaryExpression{
					Operator: Plus,
					Left:     expr,
					Right:    value,
				},
			}, nil
		case decrement:
			return &VariableAssignment{
				Operator: Assign,
				Target:   expr,
				Value: &BinaryExpression{
					Operator: Minus,
					Left:     expr,
					Right:    value,
				},
			}, nil
		default:
			return &VariableAssignment{
				Operator: Assign,
				Target:   expr,
				Value:    value,
			}, nil
		}
	}

	return expr, nil
}

func (p *parser) parseType() DeclaredType {
	if p.match(identifier) {
		switch p.previous().text {
		case "Int":
			return IntType{}
		case "Float":
			return FloatType{}
		case "Str":
			return StringType{}
		case "Bool":
			return BooleanType{}
		default:
			return CustomType{
				Name: p.previous().text,
			}
		}
	}
	return nil
}

func (p *parser) expressionStatement() (Statement, error) {
	expr, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	p.match(new_line)
	return expr, nil
}

func (p *parser) functionDef() (Statement, error) {
	if p.match(fn) {
		name := ""
		if p.check(identifier) {
			name = p.consume("identifier", "Expected function name after 'fn'").text
		}
		p.consume(left_paren, "Expected parameters list")
		params := []Parameter{}
		for !p.match(right_paren) {
			isMutable := p.match(mut)
			nameToken := p.consume(identifier, "Expected parameter name")
			p.consume(colon, "Expected ':' after parameter name")
			paramType := p.parseType()
			params = append(params, Parameter{
				Mutable: isMutable,
				Name:    nameToken.text,
				Type:    paramType,
			})
			p.match(comma)
		}
		returnType := p.parseType()

		p.consume(left_brace, "Expected function body")
		statements := []Statement{}
		for !p.check(right_brace) {
			stmt, err := p.parseStatement()
			if err != nil {
				return nil, err
			}
			statements = append(statements, stmt)
		}

		p.consume(right_brace, "Unclosed function body")

		if name == "" {
			return &AnonymousFunction{
				Parameters: params,
				ReturnType: returnType,
				Body:       statements,
			}, nil
		}

		return &FunctionDeclaration{
			Name:       name,
			Parameters: params,
			ReturnType: returnType,
			Body:       statements,
		}, nil
	}

	return p.iterRange()
}

func (p *parser) parseExpression() (Expression, error) {
	return p.functionDef()
}

func (p *parser) iterRange() (Expression, error) {
	start, err := p.or()
	if err != nil {
		return nil, err
	}

	if p.match(dot_dot) {
		end, err := p.or()
		if err != nil {
			return nil, err
		}

		return &RangeExpression{
			Start: start,
			End:   end,
		}, nil
	}

	return start, nil
}

func (p *parser) or() (Expression, error) {
	left, err := p.and()
	if err != nil {
		return nil, err
	}

	if p.match(or) {
		right, err := p.or()
		if err != nil {
			return nil, err
		}
		return &BinaryExpression{
			Operator: Or,
			Left:     left,
			Right:    right,
		}, nil
	}
	return left, nil
}

func (p *parser) and() (Expression, error) {
	left, err := p.comparison()
	if err != nil {
		return nil, err
	}
	if p.match(and) {
		right, err := p.and()
		if err != nil {
			return nil, err
		}
		return &BinaryExpression{
			Operator: And,
			Left:     left,
			Right:    right,
		}, nil
	}
	return left, nil
}

func (p *parser) comparison() (Expression, error) {
	left, err := p.modulo()
	if err != nil {
		return nil, err
	}
	for p.match(greater_than, greater_than_equal, less_than, less_than_equal, equal_equal) {
		opToken := p.previous()
		var operator Operator
		switch opToken.kind {
		case greater_than:
			operator = GreaterThan
		case greater_than_equal:
			operator = GreaterThanOrEqual
		case less_than:
			operator = LessThan
		case less_than_equal:
			operator = LessThanOrEqual
		case equal_equal:
			operator = Equal
		}
		right, err := p.modulo()
		if err != nil {
			return nil, err
		}
		left = &BinaryExpression{
			Operator: operator,
			Left:     left,
			Right:    right,
		}
	}
	return left, nil
}

func (p *parser) modulo() (Expression, error) {
	left, err := p.addition()
	if err != nil {
		return nil, err
	}
	if p.match(percent) {
		right, err := p.addition()
		if err != nil {
			return nil, err
		}
		return &BinaryExpression{
			Operator: Modulo,
			Left:     left,
			Right:    right,
		}, nil
	}
	return left, nil
}

func (p *parser) addition() (Expression, error) {
	left, err := p.multiplication()
	if err != nil {
		return nil, err
	}

	for p.match(plus, minus) {
		opToken := p.previous()
		operator := Plus
		if opToken.kind == minus {
			operator = Minus
		}
		right, err := p.multiplication()
		if err != nil {
			return nil, err
		}
		left = &BinaryExpression{
			Operator: operator,
			Left:     left,
			Right:    right,
		}
	}
	return left, nil
}

func (p *parser) multiplication() (Expression, error) {
	left, err := p.unary()
	if err != nil {
		return nil, err
	}
	for p.match(star, slash) {
		opToken := p.previous()
		operator := Multiply
		if opToken.kind == slash {
			operator = Divide
		}

		right, err := p.unary()
		if err != nil {
			return nil, err
		}
		left = &BinaryExpression{
			Operator: operator,
			Left:     left,
			Right:    right,
		}
	}
	return left, nil
}

func (p *parser) unary() (Expression, error) {
	if p.match(minus, not) {
		opToken := p.previous()
		if opToken.kind == not {
			operand, err := p.parseExpression()
			if err != nil {
				return nil, err
			}
			return &UnaryExpression{
				Operator: Not,
				Operand:  operand,
			}, nil
		} else {
			operand, err := p.unary()
			if err != nil {
				return nil, err
			}
			return &UnaryExpression{
				Operator: Minus,
				Operand:  operand,
			}, nil
		}
	}
	return p.memberAccess()
}

func (p *parser) memberAccess() (Expression, error) {
	expr, err := p.call()
	if err != nil {
		return nil, err
	}
	for p.match(dot, colon_colon) {
		if p.previous().kind == dot {
			call, err := p.call()
			if err != nil {
				return nil, err
			}

			switch prop := call.(type) {
			case *Identifier:
				expr = &InstanceProperty{
					Target:   expr,
					Property: *prop,
				}
			case *FunctionCall:
				expr = &InstanceMethod{
					Target: expr,
					Method: *prop,
				}
			}
		} else {
			call, err := p.call()
			if err != nil {
				return nil, err
			}

			switch prop := call.(type) {
			case *Identifier:
				expr = &StaticProperty{
					Target:   expr,
					Property: *prop,
				}
			case *FunctionCall:
				expr = &StaticFunction{
					Target:   expr,
					Function: *prop,
				}
			}
		}
	}
	return expr, nil
}

func (p *parser) call() (Expression, error) {
	expr, err := p.primary()
	if err != nil {
		return nil, err
	}

	// todo: to support chaining, wrap in for loop
	// ex: foo()()()
	if p.match(left_paren) {
		args := []Expression{}
		for !p.check(right_paren) {
			arg, err := p.parseExpression()
			if err != nil {
				return nil, err
			}
			args = append(args, arg)
			p.match(comma)
		}
		p.consume(right_paren, "Unclosed function call")
		return &FunctionCall{
			Name: expr.(*Identifier).Name,
			Args: args,
		}, nil
	}

	return expr, nil
}

func (p *parser) primary() (Expression, error) {
	if p.match(number) {
		return &NumLiteral{
			Value: p.previous().text,
		}, nil
	}
	if p.match(complex_string) {
		return p.interpolatedString()
	}
	if p.match(string_) {
		return &StrLiteral{
			Value: p.previous().text,
		}, nil
	}
	if p.match(true_, false_) {
		return &BoolLiteral{
			Value: p.previous().text == "true",
		}, nil
	}
	if p.match(identifier) {
		return &Identifier{
			Name: p.previous().text,
		}, nil
	}
	if p.match(left_paren) {
		expr, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		p.consume(right_paren, "Expected ')' after expression")
		return expr, nil
	}
	panic(fmt.Errorf("unmatched primary expression: %s", p.peek().kind))
}

func (p *parser) interpolatedString() (Expression, error) {
	chunks := []Expression{}
	tok := p.previous()
	for i := range tok.chunks {
		chunk := tok.chunks[i]
		if chunk.kind == string_ {
			chunks = append(chunks, &StrLiteral{
				Value: chunk.text,
			})
		} else {
			expr, err := p.parseExpression()
			if err != nil {
				return nil, err
			}
			chunks = append(chunks, expr)
		}
	}

	if len(chunks) == 1 {
		return chunks[0], nil
	}
	return &InterpolatedStr{
		Chunks: chunks,
	}, nil
}

func (p *parser) advance() token {
	if !p.isAtEnd() {
		p.index++
	}
	return p.tokens[p.index-1]
}

/* assert that the current token is the provided kind and return it */
func (p *parser) consume(kind kind, message string) token {
	if p.isAtEnd() {
		panic(fmt.Errorf("Unexpected end of input"))
	}
	if p.peek().kind == kind {
		return p.advance()
	}

	panic(fmt.Errorf("%s at line %d, column %d",
		message, p.tokens[p.index].line, p.tokens[p.index].column))
}

/* conditionally advance if the current token is the provided kind */
func (p *parser) match(kinds ...kind) bool {
	if slices.Contains(kinds, p.peek().kind) {
		p.advance()
		return true
	}
	return false
}

func (p *parser) check(kind kind) bool {
	return p.peek().kind == kind
}

func (p *parser) peek() *token {
	if p.index >= len(p.tokens) {
		return nil
	}
	return &p.tokens[p.index]
}

func (p *parser) previous() *token {
	if p.index == 0 {
		return nil
	}
	return &p.tokens[p.index-1]
}

func (p *parser) isAtEnd() bool {
	return p.tokens[p.index].kind == eof
}
